File: problem_7/java/problem7.java

public class problem7
{
  
  public static void main( String[] args)
  {
    int count = 1;
    int last = 0;

    for ( int i = 3; count < 10001; i += 2)
    {
      if (prime(i))
      {
        count += 1;
        last = i;
      }
    }

    System.out.println(last);
  }

  private static boolean prime( int divided)
  {
    int divisor = 3;
    long divided_sqrt = Math.round(Math.sqrt(divided));

    while( divided_sqrt >= divisor)
    {
      if( divided % divisor == 0)
      {
        return false;
      }

      divisor += 2;
    }

    return true;
  }
}

File: problem_7/haskell/problem7.hs
module Main where

prime_wrapper divided = prime 3 divided (round . sqrt. fromIntegral $ divided)

prime divisor divided sqrt_divided
  | divisor > sqrt_divided = [divided]
  | mod divided divisor == 0 = []
  | otherwise = next_prime
  where next_prime = prime (divisor + 2) divided sqrt_divided

main :: IO ()
main = print $ primes !! 10000
  where primes = 2 : [ x | x <- [3,5..],  prime_wrapper x /= [] ]
